File: notifications-polyglot/golang/worker/worker.go
package worker

import (
	"context"
	"encoding/json"
	"log"
	"time"

	"github.com/IBM/sarama"
	"github.com/example/notification-service-go/adapter"
	"github.com/example/notification-service-go/service"
)

type NotificationWorker struct {
	brokers       []string
	group         string
	topic         string
	statusService *service.NotificationStatusService
}

func NewNotificationWorker(brokers []string, group string, statusService *service.NotificationStatusService) *NotificationWorker {
	return &NotificationWorker{
		brokers:       brokers,
		group:         group,
		topic:         "notifications",
		statusService: statusService,
	}
}

func (w *NotificationWorker) Start() {
	config := sarama.NewConfig()
	config.Consumer.Group.Rebalance.Strategy = sarama.NewBalanceStrategyRoundRobin()
	config.Consumer.Offsets.Initial = sarama.OffsetOldest

	consumer, err := sarama.NewConsumerGroup(w.brokers, w.group, config)
	if err != nil {
		log.Fatalf("Error creating consumer group client: %v", err)
	}

	handler := &ConsumerHandler{
		statusService: w.statusService,
	}

	for {
		if err := consumer.Consume(context.Background(), []string{w.topic}, handler); err != nil {
			log.Printf("Error from consumer: %v", err)
			time.Sleep(2 * time.Second)
		}
	}
}

type ConsumerHandler struct {
	statusService *service.NotificationStatusService
}

func (ConsumerHandler) Setup(_ sarama.ConsumerGroupSession) error   { return nil }
func (ConsumerHandler) Cleanup(_ sarama.ConsumerGroupSession) error { return nil }
func (h ConsumerHandler) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
	factory := adapter.NewChannelFactory()

	for msg := range claim.Messages() {
		var req service.NotificationRequest
		if err := json.Unmarshal(msg.Value, &req); err != nil {
			log.Printf("Failed to unmarshal message: %v", err)
			sess.MarkMessage(msg, "")
			continue
		}

		id := string(msg.Key)
		log.Printf("Processing notification [%s]: sending to channels %v", id, req.Channels)

		for _, channel := range req.Channels {
			// Update status to PROCESSING
			if err := h.statusService.UpdateChannelStatus(id, channel, "PROCESSING"); err != nil {
				log.Printf("Failed to update status to PROCESSING for %s channel %s: %v", id, channel, err)
			}

			adp, err := factory.GetAdapter(channel)
			if err != nil {
				log.Printf("Error getting adapter for channel %s: %v", channel, err)
				h.statusService.UpdateChannelStatus(id, channel, "FAILED")
				continue
			}

			success := true
			for _, recipientID := range req.RecipientIDs {
				if err := adp.Send(recipientID, req.Message); err != nil {
					log.Printf("Error sending to %s via %s: %v", recipientID, channel, err)
					success = false
				}
			}

			if success {
				h.statusService.UpdateChannelStatus(id, channel, "COMPLETED")
			} else {
				h.statusService.UpdateChannelStatus(id, channel, "FAILED")
			}
		}

		log.Printf("Notification [%s] processed successfully", id)

		sess.MarkMessage(msg, "")
	}
	return nil
}

File: notifications-polyglot/rust/src/worker.rs
use rdkafka::config::ClientConfig;
use rdkafka::consumer::{Consumer, StreamConsumer};
use rdkafka::Message;
use crate::api::NotificationRequest;
use crate::adapter::ChannelFactory;
use crate::status_service::NotificationStatusService;
use std::sync::Arc;
use tracing::{info, warn, error};

pub async fn run_worker(
    brokers: &str,
    group_id: &str,
    status_service: Arc<NotificationStatusService>,
) -> anyhow::Result<()> {
    let factory = ChannelFactory::new();
    
    let consumer: StreamConsumer = ClientConfig::new()
        .set("group.id", group_id)
        .set("bootstrap.servers", brokers)
        .set("enable.partition.eof", "false")
        .set("session.timeout.ms", "6000")
        .set("enable.auto.commit", "true")
        .create()?;

    consumer.subscribe(&["notifications"])?;

    info!("Worker started, listening on topic 'notifications'");

    loop {
        match consumer.recv().await {
            Err(e) => warn!("Kafka error: {}", e),
            Ok(m) => {
                let payload = match m.payload_view::<str>() {
                    None => "",
                    Some(Ok(s)) => s,
                    Some(Err(e)) => {
                        warn!("Error while deserializing message payload: {:?}", e);
                        ""
                    }
                };
                
                let key = match m.key_view::<str>() {
                    None => "",
                    Some(Ok(s)) => s,
                    Some(Err(_)) => "",
                };

                info!("Processing notification [{}]: payload size {}", key, payload.len());

                if let Ok(req) = serde_json::from_str::<NotificationRequest>(payload) {
                    info!("Processing notification [{}]: sending to channels {:?}", key, req.channels);
                    
                    for channel in &req.channels {
                        // Update status to PROCESSING
                        if let Err(e) = status_service.update_channel_status(key, channel, "PROCESSING").await {
                            error!("Failed to update status to PROCESSING: {}", e);
                        }
                        
                        if let Some(adapter) = factory.get_adapter(channel) {
                            let mut success = true;
                            for recipient_id in &req.recipient_ids {
                                if let Err(e) = adapter.send(recipient_id, &req.message).await {
                                    error!("Failed to send to {} via {}: {}", recipient_id, channel, e);
                                    success = false;
                                }
                            }
                            
                            // Update status based on result
                            let channel_status = if success { "COMPLETED" } else { "FAILED" };
                            if let Err(e) = status_service.update_channel_status(key, channel, channel_status).await {
                                error!("Failed to update channel status: {}", e);
                            }
                        } else {
                            warn!("No adapter found for channel: {}", channel);
                            if let Err(e) = status_service.update_channel_status(key, channel, "FAILED").await {
                                error!("Failed to update status to FAILED: {}", e);
                            }
                        }
                    }
                    
                    info!("Notification [{}] processed successfully", key);
                } else {
                    error!("Failed to parse notification payload");
                }
            }
        };
    }
}
